kernel/timer.c

```diff
@@ -104,29 +103,33 @@ void k_timer_init(struct k_timer *timer,
 }
 
 
-void z_impl_k_timer_start(struct k_timer *timer, s32_t duration, s32_t period)
+void z_impl_k_timer_start(struct k_timer *timer, k_timeout_t duration,
```

Comment:
using k_timeout_t for duration does not look right, I guess this is the same case as sleep(), which sleeps for the specified duration and not up to a given timeout value. Maybe k_timeout_t is just the wrong name to be used if the implementation ends up to be the same.

Reply:
felt the same regarding k_sleep parameter. What about `k_ticks_t` isn't it a better name for that? Also in timeout context.

Reply:
`k_time_t`?

Reply:
A timeout is the abstraction for multiunit real time values.  It's true that in the case of sleep and period, the idea of an "absolute" time gets more confusing, but the behavior still works as specified (i.e. you can sleep until "516 seconds since boot", or set a k_timer to recur at "2pm tomorrow", though of course that timer will then start recurring repeatedly until you cancel it as the time is now in the past, so the utility isn't very high).

A k_ticks_t is just the type of a size needed to store system time.  It's like size_t.  It doesn't store data.

Reply:
@andyross well, sorta.  This is where it would have been better to define terms first.

In my understanding a timeout is an abstraction for a point in time at which something should occur, expressed as an absolute or relative instant based on some clock.  `k_sleep()` and the duration field of `k_timer()` both operate on timeouts.

The period field of `k_timer()` indicates an interval between repeated timeouts.  It's reasonable to require that to be a relative timeout.  It's not particularly useful to allow it to be absolute, though if it was it would have the semantics you suggest.  I particularly don't like having to pass `K_NO_WAIT` to duration in order to indicate that the timer is one-shot, though I would be content with `K_ONE_SHOT` as an alias to be used in that context.

Reply:
Honestly your definition and mind seem identical to my ears, but I'll go with yours if you like.  There's actually a whole section on timeouts in #19201 also.

FWIW: K_FOREVER works too., and the semantics in English seem to make more sense if you use that.  Is that OK, or is K_ONE_SHOT alias a new requirement for merge?  I'll update the docs to clarify.